### PR TITLE
$ functions refactoring

### PR DESCRIPTION
**Pull request by MaceWindu**
Another refactoring from #3627

- move $ functions to separate class (names and constructors)
- fix some discovered places where `CanBeNull` value were lost
- improve case/coalesce typing

### CODE REVIEW

**Review by sdanyliv — DISMISSED** [review]

File: Source/LinqToDB/SqlProvider/BasicSqlOptimizer.cs [hunk 1]
```diff
@@ -2529,9 +2569,9 @@ protected virtual ISqlExpression ConvertConvertion(SqlFunction func)
 						new SqlPredicate.ExprExpr(par, SqlPredicate.Operator.NotEqual, new SqlValue(0),
 							Configuration.Linq.CompareNullsAsValues ? false : null)));
 
-				return new SqlFunction(func.SystemType, "CASE", sc, new SqlValue(true), new SqlValue(false))
+				return new SqlFunction(func.SystemType, "CASE", false, true, sc, new SqlValue(true), new SqlValue(false))
 				{
-					CanBeNull = false,
```
Comment (viceroypenguin) [inline Source/LinqToDB/SqlProvider/BasicSqlOptimizer.cs:2534]: comma should be left on this line - reduces noise in future PRs if additional properties are added.

File: Source/LinqToDB/SqlProvider/BasicSqlOptimizer.cs [hunk 2]
```diff
@@ -2509,7 +2546,10 @@ protected virtual ISqlExpression ConvertConvertion(SqlFunction func)
 			else if (from.Type.SystemType == typeof(short) && to.Type.SystemType == typeof(int))
 				return func.Parameters[2];
 
-			return new SqlFunction(func.SystemType, "Convert", to, func.Parameters[2]);
+			return new SqlFunction(func.SystemType, "Convert", false, true, to, func.Parameters[2])
+			{
+				CanBeNull = func.Parameters[2].CanBeNull
```
Comment (viceroypenguin) [inline Source/LinqToDB/SqlProvider/BasicSqlOptimizer.cs:2551]: comma should be left on this line - reduces noise in future PRs if additional properties are added.

**Review by viceroypenguin — APPROVED** [review]

**Pull request merged**